lib/CGameState.cpp

```diff
@@ -381,8 +381,22 @@ void CGameState::CrossoverHeroesList::addHeroToBothLists(CGHeroInstance * hero)
 
 void CGameState::CrossoverHeroesList::removeHeroFromBothLists(CGHeroInstance * hero)
 {
-	heroesFromPreviousScenario -= hero;
-	heroesFromAnyPreviousScenarios -= hero;
+	for (auto obj : heroesFromPreviousScenario)
+	{
+		if (obj->name == hero->name)
+		{
+			heroesFromPreviousScenario -= obj;
+			break;
+		}
+	}
+	for (auto obj : heroesFromAnyPreviousScenarios)
+	{
+		if (obj->name == hero->name) 
```

Comment:
The hero pointer here is a result of the following searching in the `heroesFromAnyPreviousScenarios` container:
`range::find_if(crossoverHeroes.heroesFromAnyPreviousScenarios, [heroPlaceholder](CGHeroInstance * hero)`
So,  is this condition really necessary for this loop?

Reply:
If the method is called "removeFromBothList" it should search and delete from both lists, doing exact thing for both lists.
Maybe currently its callers dont need to search the second list, but method uses can change over time. And it will be difficult to next code readers to understand why we search first list, and dont search the second one.
When methods do their name, its more managable code.